### ISSUE TITLE
Make IDbContext

### ISSUE THREAD

**Issue by eltiare** [dotnet/efcore#16470]
I could not find anything specific to this particular problem. Apologies if this is a dup. I feel an interface for DbContext would be helpful for inheriting context interfaces.  Something like this: 

```
    public interface IContext : IDbContext
    {
        DbSet<Review> Reviews { get; }
        DbSet<Location> Locations { get; }
    }

    public class Context : DbContext, IContext
    {
        public DbSet<Review> Reviews { get; set; }
        public DbSet<Location> Locations { get; set; }
        
        public Context(DbContextOptions<Context> options) : base(options)
        {
        }
    }

```

I'm thinking it might make testing easier because you could drop in anything that conforms to the interface in testing / different environments. 

EF Core version: 2.2.4

**Comment by ajcvickers** [comment]
@eltiare Shipping an interface for `DbContext` creates problems for evolving the API going forward. Also, we have found that it is often better to use some form of repository pattern if you really want to expose the context services through an interface. That way you control and evolve the interface.

**Comment by douglasg14b** [comment]
@ajcvickers 

I might be completely missing the idea here, please bear with me:


> Shipping an interface for DbContext creates problems for evolving the API going forward.

Can you expand on that? Isn't the interface SUPPOSED to be the API contract?

The lack of an `IDbContext` interface is somewhat breaking to the whole dependency inversion principle? Any interface I make has no equality to what would be `IDbContext` that allows `DbContext` to be injected in it's place, I always have to refer back to my concrete implementation of `DbContext`, which breaks the clean flow of DI in some serious ways.

EF is already a repository pattern, it seems counter-intuitive to force users to implement their own abstraction on top of that if it isn't necessary? Sure, some will want to, but in other cases it's not a necessity. It feels like selecting implementation of principles that Asp.Net Core and other .Net Core projects has been putting to great use.

_____________________________________________

**Example:**

Say I have three projects. Following the Clean Architecture, but implementing DDD (Domain Driven Design):

* Core (lvl 0)
  * Defines interfaces and Dtos 
* Domain (lvl 1)
  * Defined business and domain logic
  * *Defines domain entities/models with included business logic*
  * Defines concrete Domain Service implementations (That access EF)
* Infrastructure (lvl 2)
  * Implements concrete classes of interfaces defined in Core
  * Provides DI services and configurations
  * Defines concrete Application Services from interfaces in Core

The reference structure goes `Core <- Domain <- Infrastructure`.

The necessity to have business logic within domain entities (specifically the aggregate root), puts those into the `Domain` layer. Which necessitates my `IApplicationContext` to be defined there as Core cannot reference the layer above it. This is a compromise between "Clean Architecture" and DDD purity.

My Domain Services expect an interface that's equatable to `DbContext` to be injected, the `IApplicationContext`. The concrete implementation is provided by the `Infrastructure` project, but the Domain layer doesn't need to know this, it only needs the contract.

The problem now is `IApplicationContext` has no equality to `DbContext`'s API. I could add method definitions to my interface from `DbContext`, but this is just asking for breaking changes in my concrete implementation as soon as any `DbContext` method signatures change, which would be a non-issue with a contract.

______________________________________________

The jist of this is the lack of an abstract API for `DbContext` necessitates all sorts of unnecessary workarounds as soon as you need strong DI. My example is a flawed one, but I've seen this ran into before in other use cases where strict DI is enforced...

**Comment by tb-mtg** [comment]
> Shipping an interface for DbContext creates problems for evolving the API going forward. 

Everything is evolving isn't it? This should not be a reason to not use Interfaces. 

I agree with @eltiare that IDbContext would be helpful.

Even if the API is evolving, simply update the interface if necessary to allow for any evolutions.  It will only affect those that are referencing the interface, and they can adjust the code appropriately.

An example could be:

```
  public interface IDbContext : IDisposable, IInfrastructure<IServiceProvider>, IDbContextDependencies, IDbSetCache, IDbQueryCache, IDbContextPoolable {
    ChangeTracker ChangeTracker { get; }
    DatabaseFacade Database { get; }
    EntityEntry Add(object entity);
    EntityEntry<TEntity> Add<TEntity>(TEntity entity) where TEntity : class;
    Task<EntityEntry> AddAsync(object entity, CancellationToken cancellationToken = default);
    Task<EntityEntry<TEntity>> AddAsync<TEntity>(TEntity entity, CancellationToken cancellationToken = default) where TEntity : class;
    void AddRange(IEnumerable<object> entities);
    void AddRange(params object[] entities);
    Task AddRangeAsync(IEnumerable<object> entities, CancellationToken cancellationToken = default);
    Task AddRangeAsync(params object[] entities);
    EntityEntry<TEntity> Attach<TEntity>(TEntity entity) where TEntity : class;
    EntityEntry Attach(object entity);
    void AttachRange(params object[] entities);
    void AttachRange(IEnumerable<object> entities);
    EntityEntry<TEntity> Entry<TEntity>(TEntity entity) where TEntity : class;
    EntityEntry Entry(object entity);
    bool Equals(object obj);
    object Find(Type entityType, params object[] keyValues);
    TEntity Find<TEntity>(params object[] keyValues) where TEntity : class;
    Task<TEntity> FindAsync<TEntity>(params object[] keyValues) where TEntity : class;
    Task<object> FindAsync(Type entityType, object[] keyValues, CancellationToken cancellationToken);
    Task<TEntity> FindAsync<TEntity>(object[] keyValues, CancellationToken cancellationToken) where TEntity : class;
    Task<object> FindAsync(Type entityType, params object[] keyValues);
    int GetHashCode();
    DbQuery<TQuery> Query<TQuery>() where TQuery : class;
    EntityEntry Remove(object entity);
    EntityEntry<TEntity> Remove<TEntity>(TEntity entity) where TEntity : class;
    void RemoveRange(IEnumerable<object> entities);
    void RemoveRange(params object[] entities);
    int SaveChanges(bool acceptAllChangesOnSuccess);
    int SaveChanges();
    Task<int> SaveChangesAsync(bool acceptAllChangesOnSuccess, CancellationToken cancellationToken = default);
    Task<int> SaveChangesAsync(CancellationToken cancellationToken = default);
    DbSet<TEntity> Set<TEntity>() where TEntity : class;
    string ToString();
    EntityEntry Update(object entity);
    EntityEntry<TEntity> Update<TEntity>(TEntity entity) where TEntity : class;
    void UpdateRange(params object[] entities);
    void UpdateRange(IEnumerable<object> entities);
  }
```

**Comment by ajcvickers** [comment]
The issue is that adding a new method to an interface breaks any existing implementations of that interface. However, now with interface default methods we may be able to mitigate this in ways that were not possible in the past. Re-opening to consider.

**Comment by ajcvickers** [comment]
We discussed this in triage and it isn't something we are going to implement, mainly because attempting to create different implementations of the `DbContext` API (whether for testing or other purposes) is very difficult and tends to lead people down a pit of failure. Also, we don't believe it is bad to have an interface registered in D.I. that your code depends on, but which its implementation depends on an actual `DbContext` registered in D.I.

**Comment by douglasg14b** [comment]
@ajcvickers Aren't you and your team completely sidestepping the point of an interface in this context? Asp.net core clearly shows how effectively dependency inversion and inversion of control works. EF Core on the other hand seems to pretend it doesn't exist, and in the process creates a lot of unnecessary headaches.

An `IDbContext` is not to provide an interface for users to roll their own `DbContext`, it's to provide an interface so we can effectively utilize DI. How can I have an abstract `AppContext` that can be referenced by projects BELOW where my `AppDbContext` is declarted if there is no `IDbContext` to inherit from for an `IAppContext`? You can't, an interface cannot inherit from a class.

This also ties into the lack of interfaces for `DbSet` and the lack of interface configurations for `DbSet` type parameters (Ie. `DbSet<IAuthor>`, which is then configured to instantiate/use `Author` in a configuration).

<hr>

This breaks IoC, and forces us to adopt awkward, best-practice violating architectures because EF Core is the odd one out. Unless of course I copy/paste the method & property signatures from `DbContext` into an `IDbContext`. Which now creates a maintenance problem, and should have never been necessary in the first place to support a core tenant of SOLID, a tenant that a sister framework (Asp.Net Core) widely pushes for.

**Comment by roji** [comment]
> it's to provide an interface so we can effectively utilize DI. How can I have an abstract AppContext that can be referenced by projects BELOW where my AppDbContext is declarted if there is no IDbContext to inherit from for an IAppContext? You can't, an interface cannot inherit from a class.

To understand the request better, are you asking for an interface in order to mock the DbContext for testing purposes? Or is your aim to use IDbContext in your actual product code? If it's the latter, can you provide more details on why you think an interface is needed for that, rather than subclassing DbContext? I'm looking to understand exactly what's blocking you because of the lack of an interface. An interface usually makes more sense where it's reasonable for the implementation to be swapped out for another, which really isn't the case for DbContext.

**Comment by alcjamgit** [comment]
@roji I think @douglasg14b clearly states why he wants the interface version - so we can effectively utilize DI, which means ease of testing. Yes an interface makes sense when you need implementations to be swapped out, but not usually. In fact most of the time interfaces are used to make way for testing, why do you think we make interfaces for services every time we use it in a controller or another service, do you think because we plan to swap out the implementation? Of course not, we do it for testing. Dependency injection is a pattern which is clearly promoted by the .Net community and in the case of DbContext there's a clear need for it, yet you seem to be under the impression that @douglasg14b is introducing a new concept.

Also subclassing the DbContext does not solve the issue which is again ease of testing. Yes we can introduce a repository pattern, but there's lot of reasons we don't want to do that - firstly its another abstraction over abstraction and secondly you lose a lot of power and flexibility that your ORM provides.

**Comment by roji** [comment]
@alcjamgit @douglasg14b we generally don't believe that mocking the DbContext for testing is a successful testing pattern - mainly because it implies mocking LINQ as well (in order to do any operation on that context). We generally recommend either:

* Mocking at a higher level (i.e. via a repository pattern
* Mocking at a lower level (i.e. the InMemory provider, or preferably, Sqlite in-memory mode)
* Testing against your actual production database.

**Comment by douglasg14b** [comment]
This isn't necessarily for JUST testing, don't get hungup on individual items. We **shouldn't have to enumerate the numerous ways IoC and Dependency Inversion are used in OOP languages**. There is literature on those topics spanning back decades, and more recently plenty of guidelines and documentation on how they are effectively implemented and made first-class citizens in .Net Core libraries like Asp.Net Core.

**The point is that EF Core prevents effective application of core tenants of SOLID, and it hurts the architecture of many projects, and is in direct conflict with framework design guidelines and other libraries.**

Emphasis above.

**Comment by roji** [comment]
@douglasg14b FWIW my response was about testing since this is what @alcjamgit wrote about above.

@douglasg14b I'm not aware of a principle in DI, SOLID or IoC that says that everything must have an interface. Regardless, EF Core is successfully and widely-used in many DI scenarios (e.g. virtually every ASP.NET+EF application). You are typically expected to have your AppDbContext in a relatively low data layer (I guess that would correspond to your Core layer), and reference it from above layers, wherever data operations need to happen.

**Comment by alcjamgit** [comment]
@roji, you're typically expected to have AppDbContext in a low data layer, because that's the only option (yes with InMemory Database you can directly use DbContext), but that doesn't mean you can't make it better, does it? 

What's your measure of success wide adoption? Yes, wide adoption EF Core has that, that doesn't mean developer experience is great when it comes to DI.

**Comment by eltiare** [comment]
Just in case any newbies stumble across this thread: 

Do _**NOT**_ test against your actual production database. Make a copy if you want real life data and test against that.

**Comment by roji** [comment]
@eltiare thanks, that was some poor wording on my part :smile: Corrected that comment...

**Comment by smitpatel** [comment]
If you end up testing against your actual production database, keep a backup job handy.

**Comment by zejji** [comment]
+1 for adding an IDbContext

@roji - Say for example we have an app with more than one than one type of concrete DbContext and we want to use DI and/or create methods that operate upon an interface. This may be e.g. to avoid circular references or allow for mocking. How can we do that currently? There's no IDbContext that the interfaces can inherit from, so we have to create our own which, as mentioned above, imposes an additional maintenance burden.

**Comment by douglasg14b** [comment]
I don't think @roji cares :/ they seemed unwilling to even give the principles that modern .net uses as foundations a second thought when it comes to this issue...

It would be open otherwise

**Comment by zejji** [comment]
@douglasg14b - Hopefully! It seems to be one of those omissions in EF that are glaringly obvious if you've ever stepped outside the Microsoft bubble, along with (at least until very recently) lifecycle hooks, bulk operations and filtered includes.

**Comment by roji** [comment]
@zejji and @douglasg14b, I think it's really worth distinguishing between the question of a DbContext interface for testing/mocking purposes, and the question of that interface for actual application use.

For testing ([see docs](https://docs.microsoft.com/en-us/ef/core/miscellaneous/testing/)), we do recommend testing against your actual database whenever possible. But when that's not desirable, the recommend way to mock is to use the InMemory database. That functionally gives you a DbContext working against memory data structure, which you can assert against etc. An actual mock of a DbContext for this purpose is a problematic proposal because mocking LINQ can be very difficult, depending on exactly what you're doing. However, having written all that, there's nothing preventing mocking DbContext via subclassing, rather than via an interface, if that's really desired.

Now, regarding a DbContext interface for non-testing purposes...

> Say for example we have an app with more than one than one type of concrete DbContext and we want to use DI and/or create methods that operate upon an interface.

I'm not sure I'm following your scenario. If you're proposing to inject actual DbContexts in a non-testing scenario, how do you propose to run queries against those contexts, given that LINQ queries are strongly-typed etc.? What would the shape of IDbContext look like in that scenario?

@douglasg14b I promise you that I do care - but I'm still trying to understand precisely what you find insufficient. Looking at the [clean architecture](https://docs.microsoft.com/en-us/dotnet/architecture/modern-web-apps-azure/common-web-application-architectures#clean-architecture) which you linked to above, the "ASP.NET Core Architecture" diagram represents the recommended architecture for an ASP.NET DI application. The EF DbContext is in the Infrastructure layer, the entity POCOs are in Core (which Infrastructure depends on), etc. How is the current EF design not respecting that document?

And as above, what exact API shape would IDbContext have, and how would you use it in a productive way? A code sample of what you're looking for could help clear up the confusion here.

**Comment by Rob-Page** [comment]
@roji I have run into a similar issue, and maybe I can help break it down a little better.

When working with the Clean Architecture (I have used this [template ](https://github.com/jasontaylordev/CleanArchitecture) in the past). You are essentially splitting the Product into consumable loosely coupled projects as follows.

![image](https://user-images.githubusercontent.com/6818597/96525811-2948ba00-1239-11eb-8d58-739159be244a.png)

In order to avoid a circular dependencies in the csproj configuration:

- The Application project includes the domain project(this way it can have access to the entities it is manipulating). 

- The Infrastructure project includes the Application project (in order to access the entities it is going to configure when it implements EF).

- The client layer implements both the Application and the Infrastructure layer( the infrastructure first so that the data access service can be setup and injected into the application layer).

This setup creates an awkward workaround with an interface, because the application layer cannot include the infrastructure layer(this would create a circular dependency) an IApplicationDbContext interface is created to bridge the gap. The interfaces helps infer the available EF service injected from the infrastructure layer. The Application layer cannot include the infrastructure layer in order to extend the context, resulting in the need to hardcode out all of the available context members. While this is doable, the ability to extend out an IDbContext would significantly improve the ability and accuracy of representing EF as a services when direct access to the DbContext is not avalible.

![image](https://user-images.githubusercontent.com/6818597/96527014-55196f00-123c-11eb-8df8-793b3140ae88.png)

Feel free to reference [this](https://github.com/Rob-Page/clean-architecture) repo that implements  [this](https://github.com/jasontaylordev/CleanArchitecture) template.

Files important to this issue:
- [Application.csproj](https://github.com/Rob-Page/clean-architecture/blob/master/src/Application/Application.csproj)
- [Infrastructure.csproj](https://github.com/Rob-Page/clean-architecture/blob/master/src/Infrastructure/Infrastructure.csproj)
- (client) - [WebUI.csproj](https://github.com/Rob-Page/clean-architecture/blob/master/src/WebUI/WebUI.csproj)
- (client implements Application and Infrastructure) [Startup.cs](https://github.com/Rob-Page/clean-architecture/blob/master/src/WebUI/Startup.cs)
-  [IApplicationDbContext.cs](https://github.com/Rob-Page/clean-architecture/blob/master/src/Application/Common/Interfaces/IApplicationDbContext.cs)

**Comment by roji** [comment]
@Rob-Page thanks for trying to provide more details. I've re-read all the documents and sample, and I'm still struggling to understand the issue:

As I wrote above, the ASP.NET Core Infrastructure part in the [clean architecture document](https://docs.microsoft.com/en-us/dotnet/architecture/modern-web-apps-azure/common-web-application-architectures#clean-architecture) places the DbContext in Infrastructure project, and not in the Application project. Is there any reason why EF Core should be present in your Application project, even in the form of an interface? As far as I can tell from the document, the recommendation is to have your entities there, but definitely not your context.

More importantly, seeing your IApplicationDbContext.cs, is there anything stopping you from implementing that in your concrete DbContext implementation? In other words, why is an IDbContext interface needed (which is what this issue is about), and what would its shape even be?

**Comment by Rob-Page** [comment]
@roji thank you for your response, upon further investigation and thinking through your comments, I believe that you are correct that this is not really an issue that EF Core can solve. I had knee jerk reaction to extend a an interface in order to easily access all of the methods and properties so that things, "work right." This is what erroneously drove my desire for a DbContext Interface. 

In this scenario the complex architectural setup has costs. Your last suggestion of implementing a custom IApplicationDbContext in the DbContext helped flip my brain around enough to see that this is an item that the developer should be maintaining as it is very specific to the application setup and implementation of EF Core.

I am going to test out that and a few more ideas and when I have a more concrete solution for my application I will post an update here so that people that run into this in the future will have a little direction. Thanks again for the suggestions and direction here, and to the whole team that works on EF Core, it is an amazing product!

**Comment by ajcvickers** [comment]
@Rob-Page Thanks for the additional validation on this concept/design. It's really nice to see some external validation on this thread. :-) Also, thanks for the kind words.

**Comment by tomaszzmudzinski** [comment]
@Rob-Page did't you found some clean solution for this? I am refactoring my project to use the same clean architecture as you mentioned and I am having the same problem 🤕

**Comment by readonlyden** [comment]
It's an interesting issue. `DbSet<T>` is actually a repository pattern implementation and `DbContext` is actually an implementation of the Unit Of Work pattern. 

Of course, we can create our own interfaces and implementations, maybe generic ones, but isn't this a violation of the DRY and KISS rules? Because in that we would have two repository patterns, two UoW patterns.

We can think about our UoW/Repositories as adapters to EF core ones, or bridges. It would be great to see an example of how to properly abstract from EF Core(for example, during the implementation of the DDD architecture) and not to reinvent the whole framework.

**Comment by ParadiseFallen** [comment]
@ajcvickers @roji @douglasg14b 
I create sample project that show main problem that can be easilly solved by adding IDbContext. And why you need to do it

https://github.com/ParadiseFallen/efcore-IDbContext-Issue

I hope you will understand my bad english. I will be glad to answer your questions

**Comment by honey-ramgarhia** [comment]
I ran into some similar issue which is why I end up here. Following Clean Architecture, I have these layers
1. Core: Contains DbContext. 
2. Infrastructure: Has everything provider related. Like DI extension method for registering DbContext with SqlServer, `IEntityTypeConfiguration`s.

I have no problem using DbContext as is in my Core layer for various reasons. I treat DbContext as my Repository implementation and thus Core contains DbContext and any database (provider) related things, I move them into Infrastructure layer.

But I am stuck on a little problem. I have to register `IEntityTypeConfiguration` classes and the only way I know is to add them inside `OnModelCreating` method of DbContext. But I cannot do that as Core layer cannot depend on Infrastructure. So, my question is:

- Is there any way to register these configurations inside `AddDbContext` extension method?
- Or do I had to create IDbContext interface and use it inside the core and move the actual implementation inside the Infrastructure layer? (I don't want to do this 😔)

**Comment by gsGabriel** [comment]
Guys, sorry to raise more questions about this, but I don't think creating my own interface will solve it. I have an example of this, I created my own interface in domain layer and with that I am exposing what I need to make queries and execute commands in the application layer. We know that one way to get performance is using compiled queries, the problem with this is that to use compiled queries I necessarily need the concrete class `DbContext`, so, e.g. I need to implement a repository pattern to expose compiled queries. Another example is to use the `DbContextfactory`, the factory takes the `DbContext` from DI, but only accepts concrete class. Why? if we have an interface that represents the `DbContext` in DI why not to take the new instance from that?
I don't see that creating an interface is to create my own `DbContext`, but so that I can use all the benefits of the EF without have dependency with concrete class.

**Comment by LassazVegaz** [comment]
Are there any other libraries that allow us to do Dependency Invertion?
Probably without re-creating the Repositories and Unit of Work

**Comment by StijnClaessens** [comment]
I'm also using the "clean architecture" for my solution.
I based my solution structure upon the articles of John Jacobs ([A Brief Intro to Clean Architecture, Clean DDD, and CQRS](https://softwarealchemy.net/posts/2020/2/a-brief-intro-to-clean-architecture-clean-ddd-and-cqrs) and [A Template for Clean Domain-Driven Design Architecture](https://softwarealchemy.net/posts/2020/3/a-clean-domain-driven-design-architectural-template)).
These articles are accompanied with a [project example](https://github.com/JacobsDataSolutions/OrgManager) on Github.

In the example project, you can see that an IDbContext interface is used to abstract the dbcontext in the query and command handlers (eg IApplicationWriteDbContext in AddOrUpdateCustomerCommand).
These classes are in the application layer. The application layer does not depend on the persistence layer (containing the concrete DbContext).
I don't want to depend on a concrete DbContext class in my application classes so I can accept this creation of an application IDbContext interface in my own solution.

But I have a problem when I want to refactor my query and command handlers to use DbContextFactory as per Microsofts recommendations (see [Blazor and EF Core](https://learn.microsoft.com/en-us/aspnet/core/blazor/blazor-ef-core?view=aspnetcore-10.0) or [EF Core DbContext configuration](https://learn.microsoft.com/en-gb/ef/core/dbcontext-configuration/#use-a-dbcontext-factory)). Because the IDbContextFactory depends on a concrete implementation of DbContext I can't go this recommended path.

It would be helpful if IDbContextFactory interface could work with an IDbContext interface. When creating a new DbContext it takes the concrete class that implements IDbContext from DI.
Like @gsGabriel also suggested.

**Comment by roji** [comment]
@StijnClaessens is there any reason you can't simply use the existing DbContext abstract base class, as opposed to an IDbContext which doesn't exist yet? You can similarly abstract your concrete DbContext via abstract base classes (ApplicationWriteDbContext instead of IApplicationWriteDbContext) which extend DbContext.

In general, on this thread it seems like multiple times people have requested an interface (IDbContext) although an abstract base class already exists (DbContext), but the reasons are unclear to me. In general, if you finding yourself wanting an IDbContext, please stop and try to understand exactly what the reasons are that the existing DbContext doesn't work. If you have a clear scenario for that, please post a short, succinct description including a minimal code sample, showing why you need IDbContext and DbContext wouldn't suffice.

To reiterate, there is no rule in .NET that says that "everything must have an interface", or that an interface is somehow better than an abstract base class. There are many examples of types and abstractions where there's an abstract base class rather than an interface.

**Comment by douglasg14b** [comment]
@StijnClaessens It's become the norm, at least for me, that projects with EF Core end up with a different structure that breaks away, to some degree. It's an unfortunate deciding factor in affecting dependency inversion within a project.

On projects where it's particularly troublesome, I just end up maintaining my own interface, especially when it's a preexisting project where adding EF Core would otherwise cause it's structure to have to be refactored in considerable ways. 

**Overall though, _it's usually not a problem_, but when it is it's incredibly frustrating.**

<hr>

> To reiterate, there is no rule in .NET that says that "everything must have an interface", 

The EF Core projects position appears to be the opposite in a similarly dogmatic way?

It's baffling the level of pushback against an interface here, it only makes for better DevX within a relatively common project flavor. Especially for those learning. 

Maintenance isn't a problem either. Automated validation, or even just codegen solves that problem these days. It's not like a constantly shifting surface area.

**Comment by gsGabriel** [comment]
> [@StijnClaessens](https://github.com/StijnClaessens) is there any reason you can't simply use the existing DbContext abstract base class, as opposed to an IDbContext which doesn't exist yet? You can similarly abstract your concrete DbContext via abstract base classes (ApplicationWriteDbContext instead of IApplicationWriteDbContext) which extend DbContext.
> 
> In general, on this thread it seems like multiple times people have requested an interface (IDbContext) although an abstract base class already exists (DbContext), but the reasons are unclear to me. In general, if you finding yourself wanting an IDbContext, please stop and try to understand exactly what the reasons are that the existing DbContext doesn't work. If you have a clear scenario for that, please post a short, succinct description including a minimal code sample, showing why you need IDbContext and DbContext wouldn't suffice.
> 
> To reiterate, there is no rule in .NET that says that "everything must have an interface", or that an interface is somehow better than an abstract base class. There are many examples of types and abstractions where there's an abstract base class rather than an interface.

@roji For me why the abstract class (DbContext) doesn't work: In Clean Architecture, the Core/Domain layer must define the contract (IUnitOfWork or IApplicationDbContext) but must not reference Microsoft.EntityFrameworkCore.

If I define an abstract class in Core, I cannot make it inherit from DbContext (because that requires the EF reference).

If I define a plain POCO abstract class in Core, my implementation in Infrastructure cannot inherit from it AND DbContext simultaneously due to C# Single Inheritance.

Result: Interfaces are the only structural option for the Core contract.

Why "Rolling our own Interface" breaks EF Core features: This is the critical part I mentioned earlier. Even if I create IMyDbContext in Core, I lose access to powerful EF features because of generic constraints in the framework itself.

Take IDbContextFactory<TContext> for example. The definition is:
```
public interface IDbContextFactory<TContext>
    where TContext : DbContext // <--- This constraint blocks the interface usage
{ ... }
```
I cannot inject IDbContextFactory<IMyDbContext> into my services because IMyDbContext does not inherit from DbContext.

The same applies to Compiled Queries:
```
// EF.CompileQuery expects a concrete DbContext or a type satisfying the constraint
public static Func<TContext, ...> CompileQuery<TContext, ...>(...)
    where TContext : DbContext
```
Currently, to follow strict architectural patterns, we are forced to abandon IDbContextFactory and standard Compiled Queries (or implement complex workarounds like the Repository Pattern just to hide these limitations).

We are asking for an IDbContext (or for constraints to be relaxed to class) so we can use the full power of EF Core (Factories, Pooling, Compiled Queries) without coupling our Domain layers to the concrete DbContext class.

**Comment by roji** [comment]
@gsGabriel

> In Clean Architecture, the Core/Domain layer must define the contract (IUnitOfWork or IApplicationDbContext) but must not reference Microsoft.EntityFrameworkCore.
> If I define an abstract class in Core, I cannot make it inherit from DbContext (because that requires the EF reference).

OK, though an IDbContext would suffer from the same problem; if we introduce it, it would be in Microsoft.EntityFrameworkCore - exactly like DbContext - and you wouldn't be able to reference it from your Core/Domain layer either.

In other words, just like an abstract class in Core cannot inherit DbContext (because that requires the EF reference), an interface in Core also wouldn't be able to inherit IDbContext, in case we introduce it.

> If I define a plain POCO abstract class in Core, my implementation in Infrastructure cannot inherit from it AND DbContext simultaneously due to C# Single Inheritance.

Why not have an interface in Core (as you wrote above, e.g. IUnitOfWork or IApplicationDbContext) instead of an abstract class, and then your implementation in Infrastructure would both implement that Core interface and inherit EF's DbContext? What would an IDbContext make better here?

> Why "Rolling our own Interface" breaks EF Core features: This is the critical part I mentioned earlier. Even if I create IMyDbContext in Core, I lose access to powerful EF features because of generic constraints in the framework itself.
>
> Take IDbContextFactory for example

I don't quite understand this - where is it (which layer) that you're looking to use IDbContextFactory (and compiled queries)? Since IDbContextFactory is part of EF, it cannot be used in your Core layer in any case. Whereas if you want to use IDbContextFactory in your Infrastructure layer, you have full access to your concrete DbContext implementation and can just reference that.

**Comment by roji** [comment]
@douglasg14b 

>> To reiterate, there is no rule in .NET that says that "everything must have an interface",
>
> The EF Core projects position appears to be the opposite in a similarly dogmatic way?

No, we have lots of interfaces. What we don't have is duplication where we have both an interface **and** a base class, both representing the same thing - not without a clear reason, in any case. More importantly, as with any feature or API, we need to fully understand exactly why it's needed before doing it. 

Also, in this particular case, introducing an interface like IDbContext should have a relatively high bar to justify the duplication, as it's a very basic concept in EF. Imagine a new user seeing both DbContext and IDbContext, and struggilng to understand whether they're supposed to extend the first, implement the second, or possibly both; it's a question of concept count and general API complexity. 

To summarize, we get tons of requests of the type "just do X for me". I'd love to oblige all of them, but we have to see the larger picture, understand exactly why the user is asking for it (maybe a Y would respond better to their needs), and even then, gauge whether introducing the new feature is worth the new API/concept/complexity.

**Comment by LassazVegaz** [comment]
AI suggested this approach, which is quite good
install [Microsoft.EntityFrameworkCore.Abstractions](https://www.nuget.org/packages/Microsoft.EntityFrameworkCore.Abstractions)

Create an interface for the context, like below, in the core layer

```csharp
public interface IApplicationDbContext
{
    DbSet<User> Users { get; }
    Task<int> SaveChangesAsync(CancellationToken cancellationToken = default);
}
```
The point is that the core layer still does not have to depend on an implementation. Note that you are only installing the "Asbtractions" version of EF (which I doubt if it is really an abstraction). You can generate the interface using Visual Studio easily.

What I would ask from the EF team is, would you add an option to the CLI tool to generate this interface.

**Comment by roji** [comment]
> What I would ask from the EF team is, would you add an option to the CLI tool to generate this interface.

Uhh, no, I don't think that makes sense in the EF CLI for all kinds of reasons.

**Comment by StijnClaessens** [comment]
@roji,
I can follow your reasoning about interfaces compared to (abstract) base classes. Since my IApplicationDbContext already references EF Core to access the DbSet I tried to go the abstract base class path.

In my application layer, I created an ApplicationBaseDbContext class that inherits from DbContext. My ApplicationBaseDbContext has all the needed DbSet properties so my queryHandlers (also in application layer) can access those repositories.

In my persistence layer, I can now create a concrete ApplicationDbContext class inheriting from ApplicationBaseDbContext.
Also in my persistence layer, I add the IDbContextFactory via:
services.AddDbContextFactory<ApplicationDbContext>(....)

This registers the DbContextFactory as IDbContextFactory<ApplicationDbContext>.

This makes it impossible to inject the IDbContextFactory in my queryHandlers since my application layer has no reference to the persistence layer and thus to ApplicationDbContext.

I worked around this with a wrapper ApplicationDbContextFactory in the persistence layer that implements an IApplicationDbContextFactory interface (application layer). The ApplicationDbContextFactory gets the IDbContextFactory<ApplicationDbContext> injected and returns an ApplicationBaseDbContext when creating a DbContext.
I register this ApplicationDbContextFactory as an IApplicationDbContextFactory. Now my queryHandlers get the IApplicationDbContextFactory injected.

Am I missing something while adding the DbContextFactory?
Is there no way to specify the type(s) the factory should be registered with like for AddDbContext<TContextService, TContextImplementation>()?
Something like services.AddDbContextFactory<TContextService, TContextImplementation>()?

There is only an AddDbContextFactory<TContext, TFactory>() but TFactory is constraint to IDbContextFactory<TContext>.

It works but would be nice if I could get rid of the wrapper factory.
